pkg/scheduler/plugins/reservation/transformer.go

```diff
@@ -22,210 +22,361 @@ import (
-		return nil, nil, false
-	}
-	// abort if node is not found
+func (pl *Plugin) BeforeFilter(handle frameworkext.ExtendedHandle, cycleState *framework.CycleState, pod *corev1.Pod, nodeInfo *framework.NodeInfo) (*corev1.Pod, *framework.NodeInfo, bool) {
 	node := nodeInfo.Node()
 	if node == nil {
 		return nil, nil, false
 	}
+	// Generation equals -1 means that the Reservation return logic has been executed once, but it should not happen.
+	if nodeInfo.Generation == -1 {
+		return nil, nil, false
+	}
 
-	// only continue if matchedCache is prepared in BeforePreFilter (i.e. find any reservation matched)
-	state := getPreFilterState(cycleState)
-	if state == nil {
+	state := getStateData(cycleState)
+	matchedReservationInfos := state.matched[node.Name]
+	unmatchedReservationInfos := state.unmatched[node.Name]
+	if len(matchedReservationInfos) == 0 && len(unmatchedReservationInfos) == 0 {
 		return nil, nil, false
 	}
 
-	allocatedResource, ok := state.allocatedResources[node.Name]
-	// skip when no reservation matched on this node
-	rOnNode := state.matchedCache.GetOnNode(node.Name)
-	if len(rOnNode) <= 0 && !ok {
+	nodeInfo = nodeInfo.Clone()
+	if err := pl.restoreUnmatchedReservations(cycleState, pod, nodeInfo, unmatchedReservationInfos, false); err != nil {
 		return nil, nil, false
 	}
 
-	klog.V(5).InfoS("BeforeFilter get reservation matched on node",
-		"pod", klog.KObj(pod), "node", node.Name, "count", len(rOnNode))
-	// fix-up reserved resources and ports
-	return pod, prepareFilterNodeInfo(pod, nodeInfo, rOnNode, allocatedResource), true
+	if !reservationutil.IsReservePod(pod) {
+		if err := pl.restoreMatchedReservations(cycleState, pod, nodeInfo, matchedReservationInfos, false); err != nil {
+			return nil, nil, false
+		}
+	}
+
+	return pod, nodeInfo, true
+}
+
+func (pl *Plugin) restoreUnmatchedReservations(cycleState *framework.CycleState, pod *corev1.Pod, nodeInfo *framework.NodeInfo, reservationInfos []*reservationInfo, shouldRestoreStates bool) error {
+	for _, rInfo := range reservationInfos {
+		if len(rInfo.allocated) == 0 {
+			continue
+		}
+
+		// Reservations and Pods that consume the Reservations are cumulative in resource accounting.
+		// For example, on a 32C machine, ReservationA reserves 8C, and then PodA uses ReservationA to allocate 4C,
+		// then the record on NodeInfo is that 12C is allocated. But in fact it should be calculated according to 8C,
+		// so we need to return some resources.
+		reservePod := reservationutil.NewReservePod(rInfo.reservation)
+		if err := nodeInfo.RemovePod(reservePod); err != nil {
+			klog.Errorf("Failed to remove reserve pod %v from node %v, err: %v", klog.KObj(rInfo.reservation), nodeInfo.Node().Name, err)
+			return err
+		}
+
+		for i := range reservePod.Spec.Containers {
+			reservePod.Spec.Containers[i].Resources.Requests = corev1.ResourceList{}
+		}
+		remainedResource := quotav1.SubtractWithNonNegativeResult(rInfo.allocatable, rInfo.allocated)
```

Comment:
afterPreFilter restore unmatched（将unmatched且Allocated>0的Reservation重复计算资源归还掉）为啥不直接减去Allocated
why not just subtract allocated(repeatedly counted)

Reply:
Allocated is likely to be greater than Allocatable. Because the Pod may use the remaining resources of the Reservation and the remaining resources on the node.